src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -52,107 +54,240 @@ final class P11SecretKeyFactory extends SecretKeyFactorySpi {
+        putKeyInfo(new PBKDF2KeyInfo("PBKDF2WithHmacSHA512",
+                CKP_PKCS5_PBKD2_HMAC_SHA512));
+
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA1",
+                CKM_PBA_SHA1_WITH_SHA1_HMAC, 160));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA224",
+                CKM_NSS_PKCS12_PBE_SHA224_HMAC_KEY_GEN, 224));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA256",
+                CKM_NSS_PKCS12_PBE_SHA256_HMAC_KEY_GEN, 256));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA384",
+                CKM_NSS_PKCS12_PBE_SHA384_HMAC_KEY_GEN, 384));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA512",
+                CKM_NSS_PKCS12_PBE_SHA512_HMAC_KEY_GEN, 512));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA512/224",
+                CKM_NSS_PKCS12_PBE_SHA512_HMAC_KEY_GEN, 512));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA512/256",
+                CKM_NSS_PKCS12_PBE_SHA512_HMAC_KEY_GEN, 512));
     }
 
-    // returns the PKCS11 key type of the specified algorithm
-    // no pseudo KeyTypes
     static long getPKCS11KeyType(String algorithm) {
         long kt = getKeyType(algorithm);
         if (kt == -1 || kt > PCKK_ANY) {
+            // Replace pseudo key type.
             kt = CKK_GENERIC_SECRET;
         }
         return kt;
     }
 
-    // returns direct lookup result of keyTypes using algorithm
     static long getKeyType(String algorithm) {
-        Long l = keyTypes.get(algorithm);
-        if (l == null) {
-            algorithm = algorithm.toUpperCase(Locale.ENGLISH);
-            l = keyTypes.get(algorithm);
-            if (l == null) {
-                if (algorithm.startsWith("HMAC")) {
-                    return PCKK_HMAC;
-                } else if (algorithm.startsWith("SSLMAC")) {
-                    return PCKK_SSLMAC;
-                }
-            }
-        }
-        return (l != null) ? l.longValue() : -1;
+        KeyInfo ki = getKeyInfo(algorithm);
+        return ki == null ? -1 : ki.keyType;
     }
 
     /**
      * Convert an arbitrary key of algorithm into a P11Key of provider.
      * Used in engineTranslateKey(), P11Cipher.init(), and P11Mac.init().
      */
-    static P11Key convertKey(Token token, Key key, String algo)
+    static P11Key convertKey(Token token, Key key, String svcAlgo)
             throws InvalidKeyException {
-        return convertKey(token, key, algo, null);
+        return convertKey(token, key, svcAlgo, null);
     }
 
     /**
      * Convert an arbitrary key of algorithm w/ custom attributes into a
      * P11Key of provider.
      * Used in P11KeyStore.storeSkey.
      */
-    static P11Key convertKey(Token token, Key key, String algo,
+    static P11Key convertKey(Token token, Key key, String svcAlgo,
             CK_ATTRIBUTE[] extraAttrs)
             throws InvalidKeyException {
         token.ensureValid();
-        if (key == null) {
-            throw new InvalidKeyException("Key must not be null");
-        }
```

Comment:
Is there a particular reason for removing this code block?

Reply:
Key has to be an instance of SecretKey (see [here](https://github.com/openjdk/jdk/blob/ab7ffd56bb8b93d513023d0136df55a6375c3286/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java#L268)), so a ```null``` value would lead to an InvalidKeyException anyways. We proposed to remove it in the interest of simplifying/shortening the code.